### ISSUE TITLE
Need a way to use pre-compiled DFDL schemas

### ISSUE THREAD

**Issue by mbeckerle** [apache/daffodil-vscode#615]
Large DFDL schemas can take upwards of a minute to compile.

We must have a way to specify a pre-compiled DFDL schema rather than being required to compile the schema "on the fly" at startup every time you click to debug a test. 

A parameter specifying a pre-compiled schema location is the minimum thing needed.

**Comment by scholarsmate** [comment]
Pre-compiled schema meaning the output of the save-parser subcommand (ref: https://daffodil.apache.org/cli/#save-parser-subcommand)?

**Comment by stevedlawrence** [comment]
Yeah, that's usually what we mean by a pre-compiled schema.

Note that a pre-compiled schema using that command is basically just a serialized `DataProcessor`. Maybe an alternatively is to have the extension cache DataProcessor it builds for reuse, and only rebuild it when the schema changes?

**Comment by mbeckerle** [comment]
Yeah, in the long run, having an IDE that eliminates all this schema compilation noise and just manages properly a cache of them would be great. 

Of course just making compilation of schemas way faster would be better, but that's fairly hard to do, and runtime performance is more critical than schema compilation time, IF that is, one can avoid the schema compilation time as much as possible. 

The TDML runner tries to cache compiled schemas for tests. Maybe that functionality should be exposed differently so that it can be reused from the IDE?

**Comment by lrbarber** [comment]
Does the precompiled schema have any information in it that would relate it to the .xsd file? 
How can the source schema be displayed in the debugger if we don't have the original code that was compiled?
What if there is a link, but the .xsd file has been changed since the compiled version was saved?

**Comment by arosien** [comment]
I see the basic feature as a caching problem, but I'm worried detecting transitively-included schema updates is too difficult to solve. Is there perhaps a variation of compiling a schema that only validates the syntactical structure (which handles includes)?

**Comment by arosien** [comment]
> I see the basic feature as a caching problem, but I'm worried detecting transitively-included schema updates is too difficult to solve. Is there perhaps a variation of compiling a schema that only validates the syntactical structure (which handles includes)?

Another issue mentioned a way: https://github.com/apache/daffodil-vscode/issues/76#issuecomment-1029178015

**Comment by arosien** [comment]
@mbeckerle @stevedlawrence is there a way to distinguish non-compiled schemas from pre-compiled schemas? Is a typical  file suffix used for pre-compiled schemas? The [docs](https://daffodil.apache.org/cli/#save-parser-subcommand) suggest `.xml` (vs. `.xsd` for raw schemas)... although it doesn't seem like the output is actually xml...

**Comment by stevedlawrence** [comment]
Yeah, that doc is very confusing. The saved-parser is not XML so using a .xml extension is probably a bad idea and easy to confuse with the actual schema. We should probably fix that doc.

There is not really an official standard extension, but the `packageDaffodilBin` command in the sbt-daffodil plugin gives saved parsers a name of `<format-name>-daffodilXYZ.bin`, where `XYZ` is the version of Daffodil used to build the saved parser. So we are starting to use that convention, but nothing enforces it at the moment.

One option is to look at the first handful of bytes of the file. A saved parser starts with the ascii string `DAFFODIL-<version>`, followed a NUL character, where `<version>` is the version of Daffodil the saved parser was built with. So you could look at the first handful of bytes looking for "DAFFODIL", or maybe even extract the full string up until the NUL char if that version information would be useful.

For example, `head -1z foo-daffodil380.bin` would output `DAFFODIL-3.8.0`.

**Comment by lrbarber** [comment]
@stevedlawrence If the pre-compiled parser was created from a large, complex schema that involves multiple files, is there any info that would allow different parts of the parser to be tied back to the individual component xsd files?

**Comment by stevedlawrence** [comment]
Yep, each `Parser` /`Unparser` has a `context: RuntimeData` member, and `RuntimeData` has a number of members to access schema file locations for that processor, provided by the `HasSchemaFileLocation` mixin:

https://github.com/apache/daffodil/blob/d8bc0c199d96ffd3ca9f72a63bb899aedb8fb8b8/daffodil-core/src/main/scala/org/apache/daffodil/lib/exceptions/SchemaFileLocatable.scala#L27-L34

So for to get the file/line information the startElement debug hook, you could do something like:

```scala
  override def startElement(state: PState, parser: Parser): Unit = {
    val fileLocation = parser.context.fileDescription
    val lineNumber= parser.context.lineDescription
    ...
  }
```
 
Note that depending on the Daffodil version and how the saved parser was built the file locations might be "depersonalized", meaning they are relative paths instead of absolute paths. For example older versions of Daffodil might have a fileDescription of

```
file:///home/user/path/to/dfdl/schemas/foo/src/main/resources/format/xsd/schema.dfdl.xsd
```
But newer versions will just have
```
format/xsd/schema.dfdl.xsd
```

These both can kindof lead to complications for the extension. With the former path, the saved parser might have been built on a different machine with different paths so they might not exist with whatever system you're currently debugging on. So you might have to relativeize those absolute paths somehow.  And with the later, you have to be told where the root of your source files are then scan those directories for the relative paths.

Note that modern versions of Daffodil created by the CLI or by packageDaffodilBin using the sbt-daffodil should use extension use relative paths.
